### ISSUE TITLE
Build failure on Mageia 5 with latest 1.0-rc1 version 

### ISSUE THREAD

**Issue by david-geiger** [audaspace/audaspace#1]
Hi Jörg,

I just tried to build audaspace-1.0-rc1 but without success I get these compile issues (past of full compile logs):

1- build failure with "cmake-3.0.1":
http://pastebin.com/Kg6y8X6M
2- build failure with cmake-3.3.0 and python 2 (python2-devel + python2):
http://pastebin.com/8zt7xe3K
3- build failure with cmake-3.3.0 and python 3 (python3-devel + python3):
http://pastebin.com/Jaegbc3j
4- here the spec file that I used:
http://pastebin.com/KctBUe34

Regards David.

**Comment by neXyon** [comment]
Hi David,

thanks for working on it, this helps a lot! I have attached a patch that 
fixes the cmake-3.0.1 problem and makes sure cmake also asks for python 
3, as python 2 is not supported by audaspace. I looked for solutions to 
the very last problem, the failure with cmake-3.3.0 and python 3, but 
the only possible explanaition that I found is that the python 3 
libraries are not linked properly, maybe cmake still used a different 
python version. If this error happens with the patch still, could you 
please check which python interpreter version and library versions cmake 
configures?

Thank you and best regards,
Jörg

**Comment by david-geiger** [comment]
Hi,

Where is the attached patch?

**Comment by neXyon** [comment]
Hi,

hmm, looks like github is removing the attachment. I uploaded it here: 
http://www.pasteall.org/60299/diff

Regards,
Jörg

**Comment by neXyon** [comment]
Hi again,

now I uploaded the wrong patch... Here's the correct one: 
http://www.pasteall.org/60308/diff

Sorry for the noise,
Jörg

**Comment by david-geiger** [comment]
So ok! my feedback with your patch:

- patch for cmake-3.0.1= OK
- patch for found python 3= OK
- build with cmake-3.0.1 and with only -DWITH_PYTHON:BOOL=TRUE configures= Build failure
- build with cmake-3.0.1 and with "-DWITH_PYTHON:BOOL=TRUE" + "-DWITH_PYTHON_MODULE:BOOL=FALSE" configures= Build success
- build with cmake-3.0.1 and with "-DWITH_PYTHON:BOOL=FALSE" + "-DWITH_PYTHON_MODULE:BOOL=TRUE" configures= Build success (but no libaudaspace-py.so* compiled)

- Same result with cmake 3.3.0 :)

In conclusion I have to force the "PythonInterp" bool on FALSE to build properly.

**Comment by neXyon** [comment]
Hi David,

I checked your paste again in more detail and the problem is that 
setuptools seems to execute the linking command wrong on line 253 of the 
paste:

g++ -pthread -shared -Wl,--as-needed -Wl,-z,relro -Wl,-O1 -Wl,--build-id 
-Wl,--enable-new-dtags -Wl,--as-needed -Wl,--no-undefined -Wl,-z,relro 
-Wl,-O1 -Wl,--build-id -Wl,--enable-new-dtags -Wl,--as-needed 
-Wl,--no-undefined -Wl,-z,relro -Wl,-O1 -Wl,--build-id 
-Wl,--enable-new-dtags -O2 -g -pipe -Wformat -Werror=format-security 
-Wp,-D_FORTIFY_SOURCE=2 -fstack-protector --param=ssp-buffer-size=4 
-fPIC 
build/temp.linux-x86_64-3.4/home/david/mgarepo/audaspace/BUILD/audaspace-1.0-rc1/bindings/python/PyAPI.o 
build/temp.linux-x86_64-3.4/home/david/mgarepo/audaspace/BUILD/audaspace-1.0-rc1/bindings/python/PyDevice.o 
build/temp.linux-x86_64-3.4/home/david/mgarepo/audaspace/BUILD/audaspace-1.0-rc1/bindings/python/PyHandle.o 
build/temp.linux-x86_64-3.4/home/david/mgarepo/audaspace/BUILD/audaspace-1.0-rc1/bindings/python/PySound.o 
build/temp.linux-x86_64-3.4/home/david/mgarepo/audaspace/BUILD/audaspace-1.0-rc1/bindings/python/PySequenceEntry.o 
build/temp.linux-x86_64-3.4/home/david/mgarepo/audaspace/BUILD/audaspace-1.0-rc1/bindings/python/PySequence.o 
-L. -LDebug -LRelease -laudaspace -o 
build/lib.linux-x86_64-3.4/aud.cpython-34m.so

the same line for me looks like this:

Konsole output
g++ -pthread -shared -Wl,-O1,--sort-common,--as-needed,-z,relro 
build/temp.linux-x86_64-3.4/home/nexyon/audaspace/bindings/python/PyAPI.o build/temp.linux-x86_64-3.4/home/nexyon/audaspace/bindings/python/PyDevice.o 
build/temp.linux-x86_64-3.4/home/nexyon/audaspace/bindings/python/PyHandle.o 
build/temp.linux-x86_64-3.4/home/nexyon/audaspace/bindings/python/PySound.o 
build/temp.linux-x86_64-3.4/home/nexyon/audaspace/bindings/python/PySequenceEntry.o 
build/temp.linux-x86_64-3.4/home/nexyon/audaspace/bindings/python/PySequence.o 
-L. -LDebug -LRelease -L/usr/lib -laudaspace -lpython3.4m -o 
build/lib.linux-x86_64-3.4/aud.cpython-34m.so

The important difference to notice is that my command includes the 
python library to be linked (-lpython3.4m) and yours doesn't. This is 
why you get all those linker errors of python functions not found. The 
question is why setuptools doesn't correctly add the python library to 
this line. Which version of setuptools are you using? Maybe you need to 
update python3-setuptools?

Regards,
Jörg

**Comment by david-geiger** [comment]
For now on Mageia we use python3-setuptools 3.6 version:
http://mageia.madb.org/package/show/application/0/name/python3-setuptools

**Comment by neXyon** [comment]
Uh wow, that's quite old I guess, as there is 18.1 already: 
https://pypi.python.org/pypi/setuptools

I hope it works with newer versions.

**Comment by david-geiger** [comment]
Hi Jörg,

So test build with cmake-3.3.0 and python3-setuptools-18.1 (same issue for -lpython3.4m):

http://pastebin.com/q7zUBUMs

**Comment by neXyon** [comment]
Hi David,

I dived more into this problem and found out that setuptools uses 
distutils in the background anyway. Distutils comes with python, so the 
following code is in the python3 package of Mageia.

In the file /usr/lib64/python3.4/distutils/command/build_ext.py at the 
very bottom on line 729, the variable Py_ENABLE_SHARED is checked and 
then the python library is added to the libraries. It seems like that 
Py_ENABLE_SHARED is 0 on your system for some reason, although it is 1 
in the file /usr/ Konsole output lib64/python3.4/_sysconfigdata.py in 
the original python3 package. Maybe you can find out why this is the case?

Regards,
Jörg

**Comment by pmakowski** [comment]
under mageia 5 : 

Python 3.4.3 (default, Jul  1 2015, 18:38:11) 
[GCC 4.9.2] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sysconfig
>>> sysconfig.get_config_var('Py_ENABLE_SHARED')
1

so what is the problem ?

**Comment by neXyon** [comment]
I'm on freenode in case you want a faster conversation via IRC, my nick 
there is neXyon.

As far as I can see sysconfig and distutils.sysconfig are different 
packages with same/similar functionality. Maybe you can try to add two 
print lines for debugging to the 
/usr/lib64/python3.4/distutils/command/build_ext.py file:

http://www.pasteall.org/60355/diff

See what it prints and then you can remove the two lines again.

Cheers

**Comment by neXyon** [comment]
Hi again,

I just "debugged" the problem with Dave over IRC. The problem is as follows:

In python running

__import__("distutils.sysconfig").sysconfig.get_config_var('Py_ENABLE_SHARED')

does not print anything. We further investigated and found that 
distutils.sysconfig (unlike the normal sysconfig package) tries to 
manually parse the configuration from the file

/usr/include/python3.4m/pyconfig.h

However, thanks to the "multiarch magic" in this file, the actual 
configuration is in

/usr/include/multiarch-x86_64-linux/python3.4m/pyconfig.h

and this is why distutils cannot find the configuration there. We 
hardcoded the path in

/usr/lib64/python3.4/distutils/sysconfig.py

in the get_config_h_filename() function and then building succeeded, but 
of course this is not the solution to go as it breaks any other 
architecture.

Now the debatable question is if this problem is in the way Mageia 
handles the python configuration files, or how Python's distutils reads 
the configuration, but this problem has to be solved somehow to make 
sure python extensions written in C/C++ actually compile on Mageia with 
setuptools.

There are many possible solutions: distutils could be patched in many 
positions (for the problem we are having it would be sufficient to just 
remove the last if in build_ext.py) or when creating the python package 
the pyconfig.h from multiarch-*/python*/pyconfig.h could be moved to 
python*/pyconfig.h. I am sure, you know best how to fix this problem.

Cheers,
Jörg

**Comment by david-geiger** [comment]
@pmakowski 
I tried another test simply adding the line: "#define Py_ENABLE_SHARED 1" in "/usr/include/python3.4m/pyconfig.h" file and the build is succeed.

$ cat /usr/include/python3.4m/pyconfig.h
#define _MULTIARCH_HEADER python3.4m/pyconfig.h
#include <multiarch-dispatch.h>
/* Defined if Python is built as a shared library. */
#define Py_ENABLE_SHARED 1

**Comment by pmakowski** [comment]
Thanks all for your investigation.
David, can you create an report in Mageia Bugzilla so I can find the best way to fix this multiarch issue. Thanks

**Comment by pmakowski** [comment]
I don't understand where is your problem, because in fact in setuptools build_ext.py we have : 
try:
    # Python 2.7 or >=3.2
    from sysconfig import _CONFIG_VARS
so it should have the correct information
it should never use distutils.sysconfig but sysconfig it self
and that's the first time that we reach your problem, all other python extension are building without problem
other point the /usr/include/python3.4m/pyconfig.h is correct, as it include /usr/include/multiarch-x86_64-linux/python3.4m/pyconfig.h

**Comment by david-geiger** [comment]
For reference we have an open bug report: https://bugs.mageia.org/show_bug.cgi?id=16570

**Comment by david-geiger** [comment]
Hi Jörg,

So we have fixed our cmake/distutils issue adding an 'extra_link_args = ' on setup file and it works well.

For ix86 arch:
--- audaspace-1.0/bindings/python/setup.py.in-orig	2015-08-15 00:17:00.908829159 +0200
+++ audaspace-1.0/bindings/python/setup.py.in	2015-08-15 00:19:29.371832811 +0200
@@ -25,6 +25,7 @@
                       library_dirs = ['.', 'Debug', 'Release'],
                       language = 'c++',
                       extra_compile_args = extra_args,
+                      extra_link_args = ['-L/usr/lib', '-lpython3.4m', '-lm'],
                       sources = [os.path.join(source_directory, file) for file in ['PyAPI.cpp', 'PyDevice.cpp', 'PyHandle.cpp', 'PySound.cpp', 'PySequenceEntry.cpp', 'PySequence.cpp']]
 )

For x86_64 arch:
--- audaspace-1.0/bindings/python/setup.py.in-orig	2015-08-15 00:17:00.908829159 +0200
+++ audaspace-1.0/bindings/python/setup.py.in	2015-08-15 00:17:30.837829896 +0200
@@ -25,6 +25,7 @@
                       library_dirs = ['.', 'Debug', 'Release'],
                       language = 'c++',
                       extra_compile_args = extra_args,
+                      extra_link_args = ['-L/usr/lib64', '-lpython3.4m', '-lm'],
                       sources = [os.path.join(source_directory, file) for file in ['PyAPI.cpp', 'PyDevice.cpp', 'PyHandle.cpp', 'PySound.cpp', 'PySequenceEntry.cpp', 'PySequence.cpp']]
 )

http://svnweb.mageia.org/packages?view=revision&revision=864813

**Comment by neXyon** [comment]
Hi David,

seems like this has to be the workaround until python3 distutils is fixed on Mageia, looks good! I guess everything works now with this?

Cheers

**Comment by david-geiger** [comment]
Hi Jörg,

I'm back from holiday :)

Yes with this workaround everything works now.

Regards David

**Comment by david-geiger** [comment]
Hi Jörg,

So Philippe has properly patched our python3 package and now audaspace built fine without python3 link flags on setup file.

https://bugs.mageia.org/show_bug.cgi?id=16570#c11

Time to close this issue. :)

**Comment by neXyon** [comment]
Thanks for the efforts! :)
